aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -148,17 +149,55 @@ using has_quantizer_arg =
     std::is_same<c10::intrusive_ptr<at::Quantizer>, std::decay_t<Args>>...
   >;
 
-template<class Result, class... Args>
+template <class Result, class... Args>
 struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
   static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
     TORCH_INTERNAL_ASSERT(false, "Unboxed call to a boxed kernel with unboxable parameter type at::Quantizer.");
   }
 };
 
+//
 // 2. Supported signatures, other than ref-passing.
 //
 
-template<class Result, class... Args>
+// helper class whose specializations handle single and multiple return values, respectively
+//
+template <class Result>
+struct PopResult final {
+  static Result call(Stack& stack) {
+    TORCH_CHECK(
+      stack.size() == 1,
+      "Boxed kernel was expected to push one return value to the stack, but instead pushed ",
+      stack.size(), " values."
+    );
+    return std::move(stack[0]).to<Result>();
+  }
+};
+
+template <class... Types>
+struct PopResult<std::tuple<Types...>> final {
+  using Result = std::tuple<Types...>;
+
+  static Result call(Stack& stack) {
+    // for tuple return types, boxed kernel has pushed multiple values onto the stack
+    constexpr int RetCount = sizeof...(Types);
+    TORCH_CHECK(
+      stack.size() == RetCount,
+      "Boxed kernel was expected to push ", RetCount, " return values to the stack, but instead pushed ",
+      stack.size(), " values."
+    );
+    return pop_to_tuple_impl(stack, std::make_index_sequence<RetCount>());
+  }
+private:
+  // note: this has been moved into its own helper only to avoid a parse error on `indices` otherwise.
+  // I'm sure there's an incantation that slips it past the parser but eh
```

Comment:
This way of expanding indices is a common pattern. Don't think there's an incantation to avoid the helper.

Reply:
Ugh templates are such a hack

Reply:
😬